[WIP] Verbose flag displaying progress bar for check_estimator in sklearn.utils.estimator_checks

scouvreur opened the pull request:
#### Reference Issues/PRs

Enhancement proposal suggested by @cod3licious in issue #13748. Also related to issue #11622.

#### What does this implement/fix? Explain your changes.

As suggested by @jnothman :

- [x] ~create 'verbose' parameter in check_estimator function with default value False~
- [ ] create version of check_estimator that generates tests for the pytest collector

scouvreur committed:
Created stub for check_estimator verbose flag

Created 'verbose' parameter in check_estimator function
from sklearn.utils.estimator_checks with default value
False.

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -288,7 +288,13 @@ def check_estimator(Estimator):
+        Flag to specify in order to see a progress bar.
+
     """
+    if verbose == True:
```
```suggestion
    if verbose:
```

Commit:
Update sklearn/utils/estimator_checks.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jnothman commented:
Please pursue step 2; step 1 is unnecessary, and we've correspondingly closed #13748

scouvreur commented:
Thanks @jnothman - do you have any pointers on how I could get started on step 2 ? Or any pytest resources that could point me in the right direction to generate the tests for the collector ?

scouvreur commented:
Sorry guys - is there any resource you could point me to to create tests for the pystest collector ?

amueller commented:
I think this might be a good entry point:
https://docs.pytest.org/en/latest/usage.html#calling-pytest-from-python-code

Here's how to use hooks:
https://docs.pytest.org/en/latest/example/simple.html#incremental-testing-test-steps

I think you want to work with the collection hooks?
https://docs.pytest.org/en/latest/reference.html#collection-hooks

scouvreur commented:
Thanks @amueller ! I will look into it

scouvreur committed:
Added pytest runner

Pytest runner added inside the check_estimator
function, still need to generate tests for the
collector.

On branch check_estimator_progress_verbose_mode
Your branch is up to date with 'origin/check_estimator_progress_verbose_mode'.

Changes to be committed:
	modified:   estimator_checks.py

scouvreur committed:
Merge remote-tracking branch 'upstream/master'

Merge remote-tracking branch 'upstream/master'
into check_estimator_progress_verbose_mode.

scouvreur committed:
Merge branch 'check_estimator_progress_verbose'

Merge branch 'check_estimator_progress_verbose' of
github.com:scouvreur/scikit-learn into
check_estimator_progress_verbose_mode.

jnothman commented:
I wouldn't think we want to make use of `pytest.run` here.

I think we already have a usable solution internal to scikit-learn here: https://github.com/scikit-learn/scikit-learn/blob/197f448eedb13fd267e3cc0c2a3b98c87706106d/sklearn/tests/test_common.py#L84-L113

but at the moment it relies on the internal `_yield_all_checks`. If we made that public, we could provide a pytest hook that would improve the syntax and allow for that `test_estimators` to be reduced to 

```py
@pytest.mark.estimator_checks(_tested_estimators())
def test_estimators(estimator, check):
    # Common tests for estimator instances
    with ignore_warnings(category=(DeprecationWarning, ConvergenceWarning,
                                   UserWarning, FutureWarning)):
        set_checking_parameters(estimator)
        name = estimator.__class__.__name__
        check(name, estimator)
```

or something even more succinct, such as not needing the `pytest.mark` syntax.

This assumes that it's fine to construct the estimator and identify which checks it qualifies at collection time (we currently do this in scikit-learn). This may not be efficient. To avoid constructing the estimator at collection time, you'd need to generate all the checks and skip those combinations that are inapplicable at test (not collection) time. Here one option would be to create an `estimator_check` fixture that is itself [parametrized](https://docs.pytest.org/en/latest/fixture.html#fixture-parametrize) to consider each check a different collected item.

I'm happy with the former approach - a public alternative to our current use of `_generate_checks_per_estimator` - for now.

scouvreur commented:
Thanks for the feedback @jnothman - I will work on that and update you with progress !

scouvreur committed:
Made _yield_all_checks public

Changed _yield_all_checks to a public
method and changed references to its
private version.

scouvreur committed:
Removed pytest import

scouvreur committed:
Updated import statement

amueller commented:
@jnothman I guess I wanted it to be a function, your solution would require the users to write a test, right? Which might be more sensible but not be a direct replacement for ``check_estimator``.

jnothman commented:
> I guess I wanted it to be a function, your solution would require the users to write a test, right? Which might be more sensible but not be a direct replacement for check_estimator.

check_estimator needs to be called in a test for library developers already, so I don't see what's changed. We could add magic like

```
test_my_estimator = make_check_estimator_test(MyEstimator())
```
but I don't really see the benefit.

jnothman commented:
If you really want to follow the approach I suggested above (not sure if it's the right thing to do), you will need to implement something much like https://github.com/pytest-dev/pytest/blob/2c402f4bd9cff2c6faeccb86a97364e1fa122a16/src/_pytest/python.py#L119-L128

jnothman commented:
but actually that is probably a nuisance since scikit-learn would have to provide a pytest plugin, which seems a bit silly? Maybe better off just having users do:
```
@pytest.mark.parametrize('check', yield_all_checks)
@pytest.mark.parametrize('estimator', _tested_estimators())
def test_estimators(estimator, check):
    # Common tests for estimator instances
    with ignore_warnings(category=(DeprecationWarning, ConvergenceWarning,
                                   UserWarning, FutureWarning)):
        set_checking_parameters(estimator)
        name = estimator.__class__.__name__
        check(name, estimator)
```

rth commented:
> ```
> @pytest.mark.parametrize('check', yield_all_checks)
> @pytest.mark.parametrize('estimator', _tested_estimators())
> ```
The limitation of this is that checks are estimator dependent, and having lots of skipped tests that will never run is also not ideal.

The solution proposed in https://github.com/scikit-learn/scikit-learn/pull/13843#issuecomment-503560404 sounds reasonable. It's fairly close to the one proposed in  https://github.com/scikit-learn/scikit-learn/issues/11622#issuecomment-405960168 

```py

ALL_ESTIMATORS = [Estimator1, Estimator2, etc]

@pytest.mark.parametrize(
    'check, name, estimator',
    itertools.chain.from_iterable(
         check_estimator(Estimator, evaluate=False)
         for estimator in ALL_ESTIMATORS
    )
)

(maybe with an additional helper function), where `check_estimator(Estimator, evaluate=False)` yields checks for that estimator.

jnothman commented:
Right. Thanks for the reminder that we are trying to not overwhelm it with skips.

amueller commented:
Fair, let's go with something like that. Seems pretty straight-forward, right?

jnothman commented:
I think we should have generate_only=True rather than evaluate=False, but yes, it's a good solution as far as I'm concerned.

scouvreur committed:
Merge remote-tracking branch 'upstream/master' into check_estimator_progress_verbose_mode

scouvreur committed:
Merge remote-tracking branch 'upstream/master' into check_estimator_progress_verbose_mode

scouvreur committed:
Added Evaluate flag to check_estimators()

Added bool flag for check_estimators() and
updated pytest decorator for test_estimators().

Changes to be committed:
	modified:   tests/test_common.py
	modified:   utils/estimator_checks.py

amueller commented:
@scouvreur I think we're going with #14381 right now. Not sure if there has been some miscommunication anywhere? cc @thomasjpfan

thomasjpfan commented:
Yes, it seems we are leaning toward #14381. Thank you for working on this issue @scouvreur !

scouvreur commented:
Ah I see @amueller - should I close this PR then ?

jnothman commented:
It's been closed. Thanks a lot for helping us nut this out.
